flink-connectors/flink-orc/src/test/java/org/apache/flink/orc/OrcTableSourceITCase.java

```diff
@@ -0,0 +1,289 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.orc;
+
+import org.apache.flink.api.java.DataSet;
+import org.apache.flink.api.java.ExecutionEnvironment;
+import org.apache.flink.api.java.io.DiscardingOutputFormat;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.TableEnvironment;
+import org.apache.flink.table.api.java.BatchTableEnvironment;
+import org.apache.flink.test.util.MultipleProgramsTestBase;
+import org.apache.flink.types.Row;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.List;
+
+/**
+ * Tests both {@link OrcTableSource} and {@link RowOrcInputFormat}
+ * for Table API.
+ */
+public class OrcTableSourceITCase extends MultipleProgramsTestBase {
+
+	public OrcTableSourceITCase() {
+		super(TestExecutionMode.CLUSTER);
+	}
+
+	private static String orcPath = "/Users/usmanyounas/Desktop/Lib/orc/examples/";
+
+	@Test
+	public void testOrcNestedData() throws Exception {
+
+		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+		BatchTableEnvironment tEnv = TableEnvironment.getTableEnvironment(env);
+
+		String path = orcPath + "TestOrcFile.test1.orc";
+		OrcTableSource orc = new OrcTableSource(
+			path,
+			"struct<boolean1:boolean,byte1:tinyint,short1:smallint,int1:int,long1:bigint,float1:float,double1:double,bytes1:binary,string1:string,middle:struct<list:array<struct<int1:int,string1:string>>>,list:array<struct<int1:int,string1:string>>,map:map<string,struct<int1:int,string1:string>>>"
+		);
+
+		tEnv.registerTableSource("orcTable", orc);
+
+		String query = "Select middle, list, map as mymapper from orcTable";
+		Table t = tEnv.sql(query);
+
+		DataSet<Row> dataSet = tEnv.toDataSet(t, Row.class);
+		List<Row> records = dataSet.collect();
+
+		Assert.assertEquals(records.get(0).toString(), "[1,bye, 2,sigh],[3,good, 4,bad],null");
```

Comment:
we should always check for the correct number of records and a few selected rows. However, rows might not be in the correct order if we have multiple splits.